### PR TITLE
pillar: Improve parsing of environment variables

### PR DESCRIPTION
**Pull request by rene**
This PR improves the parsing of environment variables for Edge Applications received from the controller. The go-envparse library is used to detect syntax errors that would break the shell of the shim VM (in case of containers) and to improve error messages. The checks are done in the domain manager level.

### CODE REVIEW

File: pkg/pillar/cmd/domainmgr/domainmgr.go [hunk 1]
```diff
@@ -2594,7 +2596,23 @@ func parseEnvVariablesFromCloudInit(envPairs []string) (map[string]string, error
...
+		key := strings.Trim(pair[0], " ")
+		val := strings.Trim(pair[1], " ")
+
+		// Parse the variable declaration to search for any syntax error
+		// that would break the processing shell
+		expline := key + "=" + val
+		_, err := envp.Parse(bytes.NewReader([]byte(expline)))
+		if err != nil {
+			errStr := fmt.Sprintf("Variable \"%s\" not defined properly: %s", key, err)
+			return nil, errors.New(errStr)
```
Comment (christoph-zededa) [inline pkg/pillar/cmd/domainmgr/domainmgr.go:2612]: no need to first create a string and then convert it into an `
error`
```suggestion
			return nil, fmt.Errorf("Variable \"%s\" not defined properly: %s", key, err)
```

File: pkg/pillar/cmd/domainmgr/domainmgr.go [hunk 2]
```diff
@@ -2594,7 +2596,22 @@ func parseEnvVariablesFromCloudInit(envPairs []string) (map[string]string, error
 			errStr := fmt.Sprintf("Variable \"%s\" not defined properly\nKey value pair should be delimited by \"=\"", pair[0])
 			return nil, errors.New(errStr)
 		}
-		envList[pair[0]] = pair[1]
+
```
Comment (christoph-zededa) [inline pkg/pillar/cmd/domainmgr/domainmgr.go:2599]: IMHO it would be good to explain that you want to fail parsing (while go-envparse is trying to correct), because otherwise you could just do:
```
return envp.Parse(bytes.NewReader([]byte(strings.Join(envPairs, "\n")))
```

**Comment by rene** [comment]
Converted to draft because I want to test an alternative approach with go-envparse.

**Comment by rene** [comment]
Updates in this PR:
- Use go-envparse to parse the whole environment contents and do the trimming of values

Kudos to @christoph-zededa for the suggestion.

**Review by christoph-zededa — APPROVED** [review]

File: pkg/pillar/containerd/oci_test.go [hunk 3]
```diff
@@ -34,7 +34,7 @@ const loaderRuntimeSpec = `
             "/init.sh"
         ],
         "env": [
-            "PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin"
```
Comment (eriknordmark) [inline pkg/pillar/containerd/oci_test.go:37]: Is this change required? If so is it required to quote all values e.g.,
FOO=1
needs to be changed to
FOO="1"

In any case, can't think break existing cloud-init/user-data for deployed containers?

Comment (rouming) [reply pkg/pillar/containerd/oci_test.go:37]: > In any case, can't think break existing cloud-init/user-data for deployed containers?

This is very valid concern: nothing can be worse than breaking compatibility with existing user data. 
To be on the safe side I can propose a small improvement to the existent change: here in this [line](https://github.com/lf-edge/eve/blob/7ad0126475f6ea698166ba4fbf4c1c3889d2554a/pkg/pillar/cmd/domainmgr/domainmgr.go#L1524) we compare old cloud-init version and a new one. The idea is that we can apply a new parsing mechanism only if versions differ. This will require a bit more of coding exercises (for example moving old and new versions retrieving  from the `doActivate` to the `configToStatus`), but that should not be a big deal. 

Also I would like @europaul to be involved in the review process (at least, Paul, please be aware of proposed changes, since you were the last one who touched the environment parsing logic).

**Comment by rene** [comment]
Updates in this PR:

Took into account all comments from reviewers. Thanks @eriknordmark and @rouming for bringing the important point about the backward compatibility. These are the changes now:

- All (key,value) pairs are trimmed off (spaces and double quotes), so we avoid problems like: `MYVAR=""MESS""`
- On a default shell, a declaration like this: `export "MYVAR=VALUE"` it works, so they are now allowed by pillar
- Double quotes are removed from the values by go-envparse as well, so we put them again in oci.go, as it was working before
- Since the final outcome is the same (values double quoted), all tests are equivalent and don't require any changes
- go-envparse improves not only the parsing but the error messages as well, now we have a clear indication about the error when some miss declaration is found, and this error is propagated to the cloud since it happens in domain manager level

**Review by eriknordmark — APPROVED** [review]
LGTM

File: pkg/pillar/containerd/oci_test.go (hunk 3 above)
Comment (europaul) [reply pkg/pillar/containerd/oci_test.go]: I took the test data from envparse's own [tests](https://github.com/hashicorp/go-envparse/blob/main/envparse_test.go) and ran them through the old parsing code + a `sh` process to see if we get any "false positives" - inputs that would cause an error when parsed by `envparse` but would not break the shell and the old parsing. This was the result:
```bash
"a b c=x"
"FOO=\"\\a\""
"FOO=\"\\"
"FOO=\"\\u12\""
"FOO=\"\\uabcZ\""
"FOO=\"abc \\uD83D\""
"FOO=\"abc \\uD83D \\uDE01\""
"FOO=\"abc \\uD83DDE01\""
"FOO=\"abc \\uD83D\\uDE0\""
```

That means that backward compatibility is broken on those kinds of env inputs. Even if we discard the weird unicode cases, the first three lines are still valid.

Here is the [test code](https://go.dev/play/p/DyTtRx1W0qT). It's not identical to the old and new versions, so I will run some manual tests with the full code base.

**Review by eriknordmark — APPROVED** [review]
Run eden again

File: pkg/pillar/containerd/oci_test.go (hunk 3 above)
Comment (europaul) [reply pkg/pillar/containerd/oci_test.go]: The manual tests showed that the inputs
```
"a b c=x"
"FOO=\"\\"
"FOO=\"\\u12\""
```
are accepted by the old version. The others are caught fine by the change introduced in https://github.com/lf-edge/eve/pull/3752

Comment (rene) [reply pkg/pillar/containerd/oci_test.go]: Thanks @europaul for the tests. Indeed, we cannot complain the new parser it's 100% backward compatible with the old one. However, the tests you mentioned are weird and should not be seen in a regular use case. In other words, they are correctly treated as wrong declarations by go-envparse, so let's keep the new parser (already merged) as it is... Nevertheless, it's really valid raise these points, thanks again.
